kernel/src/snapshot.rs

```diff
@@ -387,29 +393,22 @@ pub(crate) struct LastCheckpointHint {
     pub(crate) checksum: Option<String>,
 }
 
-/// Try reading the `_last_checkpoint` file.
+/// Try reading the `_last_checkpoint` file. Note that this file is optional and may not exist -
+/// causing this function to return an Err.
 ///
 /// Note that we typically want to ignore a missing/invalid `_last_checkpoint` file without failing
-/// the read. Thus, the semantics of this function are to return `None` if the file is not found or
-/// is invalid JSON. Unexpected/unrecoverable errors are returned as `Err` case and are assumed to
-/// cause failure.
-///
-/// TODO: java kernel retries three times before failing, should we do the same?
+/// the read. Thus, callers should likely not just propagate the error, but instead log a warning.
 fn read_last_checkpoint(
     storage: &dyn StorageHandler,
     log_root: &Url,
-) -> DeltaResult<Option<LastCheckpointHint>> {
+) -> DeltaResult<LastCheckpointHint> {
     let file_path = log_root.join(LAST_CHECKPOINT_FILE_NAME)?;
-    match storage
-        .read_files(vec![(file_path, None)])
-        .and_then(|mut data| data.next().expect("read_files should return one file"))
-    {
-        Ok(data) => Ok(serde_json::from_slice(&data)
-            .inspect_err(|e| warn!("invalid _last_checkpoint JSON: {e}"))
-            .ok()),
-        Err(Error::FileNotFound(_)) => Ok(None),
-        Err(err) => Err(err),
-    }
+    let last_checkpoint_data = storage
+        .read_files(vec![(file_path, None)])?
+        .next()
+        .ok_or_else(|| Error::generic("_last_checkpoint file is empty"))?;
```

Comment:
I don't think it's that the file is necessarily empty right? We should propagate the error. Assuming we have an into just:
```suggestion
        .read_files(vec![(file_path, None)])?
        .next()?;
```

Reply:
Ah this is to convert the `Option` to an error in case read_files returns an empty iterator

Reply:
Ohh hah, right, got it